source/extensions/transport_sockets/alts/tsi_socket.cc

```diff
@@ -165,10 +170,60 @@ Network::PostIoAction TsiSocket::doHandshakeNextDone(NextResultPtr&& next_result
   return Network::PostIoAction::KeepOpen;
 }
 
+void TsiSocket::onReadBufferLowWatermark() {
+  if (callbacks_->connection().state() == Network::Connection::State::Open) {
+    callbacks_->connection().readDisable(false);
+  }
+}
+
+void TsiSocket::onReadBufferHighWatermark() {
+  if (callbacks_->connection().state() == Network::Connection::State::Open) {
+    callbacks_->connection().readDisable(true);
+  }
+}
+
+Network::IoResult TsiSocket::repeatReadAndUnprotect(Buffer::Instance& buffer,
+                                                    Network::IoResult prev_result) {
+  Network::IoResult result = prev_result;
+  uint64_t total_bytes_processed = prev_result.bytes_processed_;
+  do {
+    // Check if buffer needs to be drained.
+    if (callbacks_->shouldDrainReadBuffer()) {
+      callbacks_->setTransportSocketIsReadable();
+      break;
+    } else {
+      // Do another read and unprotect.
+      if (!end_stream_read_ && !read_error_) {
+        result = raw_buffer_socket_->doRead(read_buffer_);
+        end_stream_read_ = result.end_stream_read_;
+        read_error_ = result.action_ == Network::PostIoAction::Close;
+        // No data is read or read error happens.
+        if (result.bytes_processed_ == 0) {
+          break;
+        }
+        // do unprotect only if some data is read.
+        if (result.bytes_processed_ > 0) {
+          uint64_t read_size = read_buffer_.length();
+          tsi_result status = frame_protector_->unprotect(read_buffer_, buffer);
+          ENVOY_CONN_LOG(debug, "TSI: unprotected buffer left: {} result: {}",
+                         callbacks_->connection(), read_buffer_.length(),
+                         tsi_result_to_string(status));
+          total_bytes_processed += read_size - read_buffer_.length();
+        }
+      } else {
+        break;
+      }
+    }
+  } while (true);
+
+  result.bytes_processed_ = total_bytes_processed;
+  return result;
+}
+
 Network::IoResult TsiSocket::doRead(Buffer::Instance& buffer) {
   Network::IoResult result = {Network::PostIoAction::KeepOpen, 0, false};
   if (!end_stream_read_ && !read_error_) {
-    result = raw_buffer_socket_->doRead(raw_read_buffer_);
+    result = raw_buffer_socket_->doRead(read_buffer_);
```

Comment:
I think that we may need to read in a loop while handling the handshake now that we are setting a lower initial watermark limit on the buffer.  Without it, resumption may not happen correctly for very large handshake messages.

I think that the current implementation of watermark callbacks on the read_buffer_ ends up hiding the large handshake resumption failure as calling readDisable(true) followed by readDisable(false) effectively re-registers the socket for events.

Reply:
Or how about we create another read_buffer that does not have any watermark set and use it just for the handshake? I think it will be much simpler.

Reply:
The issue is with the transition from handshake to operations past the end of the handshake.  By allowing unbounded reads during the handshake we could end up having to copy a lot of post handshake protected bytes when handling handshake complete back to the raw buffer because those bytes were read into the unbounded buffer used during the handshake.

We need to run in a loop during both the handshake and post-handshake.  Using a buffer without bounds for the handshake doesn't simplify much.

Reply:
I do not think "unbounded reads" will happen during the handshake because we are enforcing the read limit on the read_buffer (which will contain all handshake messages and unused protected bytes received from peer) via shouldDrainReadBuffer(). It then passes the data in read_buffer into the raw buffer ([here](https://github.com/envoyproxy/envoy/blob/main/source/extensions/transport_sockets/alts/tsi_socket.cc#L71)), which although does not have any read limit, thanks to the limit set on the read_buffer, the raw buffer will not hold unlimitedly large data.